### ISSUE TITLE
Lock Statement

### ISSUE THREAD

**Issue by ericrtodd** [dotnet/csharplang#750]
Currently, the lock statement is syntactic sugar for Monitor.Enter(obj) ... Monitor.Exit(obj). There are times when it would be preferable to have the Monitor.TryEnter(obj) behavior instead (skip if the lock cannot be acquired). Could there be a "trylock" type of statement that might also take a timeout value?

e.g.

```
trylock(obj) {}

trylock(obj, 2000)
{
}
```

**Comment by jjvanzon** [comment]
I am guessing the thumbs down emoticon means: just use Monitor in specialized cases. But if it is really just a monitor under the hood, this gives me another idea. I noticed the lock statement will not work in async methods, probably because a Monitor is thread-affine. Currently I work around it by using a Semaphore with max count 1. Can't the C# compiled do that for me under the hood?

**Comment by CyrusNajmabadi** [comment]
>  I noticed the lock statement will not work in async methods

Locks are completely allowed in async methods.  What is not allowed is performing an *await* inside a lock.  (You're welcome to perform them outside the lock without any issue).  The reason for this is that it's almost certainly broken as you may come back on another thread.  And another thread cannot release the monitor acquired by a different thread.

--

> Currently I work around it by using a Semaphore with max count 1. Can't the C# compiler do that for me under the hood?

First, i can't imagine how this would work.  The way lock works is by being given an actual object to lock.  i.e. you write "lock (obj)" and the compiler converts that to "Monitor.Enter(obj)".  How would this be changed by the compiler to use a Semaphore?  There is no instantiated semaphore anywhere.  What semaphore instance would the compiler use?  Other code elsewhere might also be locking on that object (including in different assemblies), yoru code would now run concurrently with it.

Second, Semaphores and monitors have different semantics.  A monitor can be entered as many times as you want, from the same owner.  A semaphore (with maxCount: 1) cannot.  Silently changing from lock to semaphore means we would break code if you had a lock in synchronous code and changed it to async.  

--

Basically, i have no idea how we would accomplish what you wanted.  I agree that it can be unpleasant having to use synchronization mechanism in async code.  But that's a good thing in my opinion.  You do need ot think very carefully about what it means given that your code is now potentially long running and bouncing across different threads.

**Comment by CyrusNajmabadi** [comment]
> There are times when it would be preferable to have the Monitor.TryEnter(obj) behavior instead (skip if the lock cannot be acquired).

I don't think this is a good idea.  Generally when TryEnter is used, and it fails, you need to have another codepath to execute to decide what to do at that point.  You don't just skip past the code you would have normally run.

**Comment by CyrusNajmabadi** [comment]
Closing until there is a full proposal provided that gives more detail and addresses tehse issues.

**Comment by jnm2** [comment]
@CyrusNajmabadi But... analyzers!
\</humor>
